packages/trader/src/App/Components/Elements/VideoStream/video-stream.tsx

```diff
@@ -1,75 +1,299 @@
+
+        const new_width = calculateNewWidth(e);
+        progress_bar_ref.current.style.setProperty('transition', 'none');
+        progress_bar_ref.current.style.setProperty('width', `${new_width}%`);
+        video_ref.current.currentTime = (Number(video_ref.current.duration) * new_width) / 100;
+
+        setIsPlaying(!video_ref.current.paused);
+    };
+
+    const dragEndHandler = (e: React.MouseEvent<HTMLElement, MouseEvent> | React.TouchEvent<HTMLElement>) => {
+        if (e.type === 'mouseup') e.preventDefault();
+        e.stopPropagation();
+
+        if (!is_dragging.current) return;
+
+        video_ref?.current?.play();
+        setIsPlaying(true);
+        setIsEnded(false);
+
+        is_dragging.current = false;
+
+        if (is_mobile) {
+            const point = document.querySelector('.player__progress-dot') as HTMLElement;
+
+            point.style.width = '1.2rem';
+            point.style.height = '1.2rem';
+            point.style.bottom = '-0.3rem';
+        }
+    };
+
+    const onRewind = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
+        e.preventDefault();
+        e.stopPropagation();
+
+        if ((e.target as HTMLElement).className === 'player__progress-dot') return;
+        if (!video_ref?.current || !progress_bar_ref?.current) return;
+
+        video_ref.current.pause();
+
+        const new_width = calculateNewWidth(e);
+
+        progress_bar_ref.current.style.setProperty('transition', 'none');
+        progress_bar_ref.current.style.setProperty('width', `${new_width}%`);
+        video_ref.current.currentTime = (Number(video_ref.current.duration) * new_width) / 100;
+
+        timer_ref.current = setTimeout(() => {
+            video_ref?.current?.play();
+            setIsPlaying(true);
+            setIsEnded(false);
+        }, 500);
+    };
+
+    const onLoadedMetadata = () => {
+        if (!video_ref?.current || !progress_bar_ref?.current) return;
+
+        setVideoDuration(video_ref.current.duration);
+        setIsPlaying(!!should_autoplay);
+        video_ref.current.volume = 0.5;
+    };
+
+    const onEnded = () => {
+        setIsPlaying(false);
+        setIsEnded(true);
+
+        // If user switch to another tab until the end of the video, the requestAnimationFrame will stop working, but we need to change progress bar to 100%
+        progress_bar_ref.current && progress_bar_ref.current.style.setProperty('width', '100%');
+    };
+
+    const repeat = React.useCallback(() => {
+        if (!video_ref?.current || !progress_bar_ref?.current) return;
+
+        setCurrentTime(video_ref.current.currentTime);
+        progress_bar_ref.current.style.setProperty('transition', 'all 0.3s linear');
+        const new_width = (video_ref.current.currentTime / Number(video_ref.current.duration)) * 100;
+        progress_bar_ref.current.style.setProperty('width', `${new_width >= 99 ? 100 : new_width}%`);
+
+        animation_ref.current = requestAnimationFrame(repeat);
```

Comment:
Please elaborate how do we benefit from requestAnimationFrame? Why do we not use setTimeout and CSS animation?

Reply:
1) setTimeout is not accurate and it is not working properly when user switches to another tab. requestAnimationFrame() calls are paused in most browsers when running in background tabs or hidden in order to improve performance and battery life.
2) requestAnimationFrame() method requests the browser to call a user-supplied callback function before the next repaint and the frequency of calls to the callback function will generally match the display refresh rate (which makes the animation smooth). So, basically, the animation is happening, when the browser is ready to repaint the page